### PR TITLE
Service id change bug

### LINKED ISSUE #203: Changing a calendar's service_id causes stop_times to get deleted
## Observed behavior

While editing this GTFS file ([gtfs_GL.zip](https://github.com/conveyal/gtfs-lib/files/2896582/gtfs_GL.zip)) in the datatools-ui editor, I changed the service_id on the calendar entry. Then, I checked some stop times in the timetable and noticed that they are completely missing for the even numbered trips. I have even verified that they are gone from the database.

## Expected behavior

Stop times should not be deleted when changing the service_id.

## Steps to reproduce the problem

See description.

## Notes on configuration and files required to reproduce the issue

gtfs-lib: https://github.com/conveyal/gtfs-lib/commit/5ec52854ac2c39dd94f8ac9d0f0fc051cf53d312
datatools-server: https://github.com/conveyal/datatools-server/commit/5943dfbe9e3acee0393c1e745a82f923a2483d2e
datatools-ui: https://github.com/conveyal/datatools-ui/commit/0f10fa25468e43b489942d4eb777706449dd6653

### PR DESCRIPTION
**Pull request by landonreed**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR fixes #203 by clarifying the conditional check in `JdbcTableWriter#updateReferencingTables` that is intended to delete stop_times and frequencies when a pattern is being deleted. The conditional was too general, so when calendar#service_id was being changed, stop_times and frequencies were at risk of deletion as well.

### CODE REVIEW

File: src/test/resources/real-world-gtfs-feeds/package-info.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/**
+ * This package contains real-life GTFS feeds. While it's best to reduce data to the simplest form so problems can be
+ * isolated more easily, adding these feeds can be very useful for creating test cases that prove bugs exist in certain
+ * edge cases.
+ */
+
+package com.conveyal.gtfs.dto;
```
Comment (evansiroky) [inline src/test/resources/real-world-gtfs-feeds/package-info.java]: Looks like some bad copy-pasta here on my part. I'll correct it.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 2]
```diff
@@ -1347,21 +1340,27 @@ private static void updateReferencingTables(
...
+                    ) {
+                        // If deleting a pattern, cascade delete stop times and frequencies for trips first. This must
+                        // happen before trips are deleted in the block below. Otherwise, the queries to select
+                        // stop_times and frequencies to delete would fail because there would be no trip records to join
+                        // with.
                         String stopTimesTable = String.join(".", namespace, "stop_times");
                         String frequenciesTable = String.join(".", namespace, "frequencies");
                         String tripsTable = String.join(".", namespace, "trips");
-                        // Delete stop times and frequencies for trips for pattern
+                        // Delete stop times and frequencies for trips for pattern FIXME SQL injection
```
Comment (evansiroky) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Please make a prepared statement to avoid the possibility for SQL injection.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 3]
```diff
@@ -1347,21 +1340,27 @@ private static void updateReferencingTables(
...
                         String deleteStopTimes = String.format(
                                 "delete from %s using %s where %s.trip_id = %s.trip_id and %s.pattern_id = '%s'",
                                 stopTimesTable, tripsTable, stopTimesTable, tripsTable, tripsTable, keyValue);
                         LOG.info(deleteStopTimes);
                         PreparedStatement deleteStopTimesStatement = connection.prepareStatement(deleteStopTimes);
                         int deletedStopTimes = deleteStopTimesStatement.executeUpdate();
                         LOG.info("Deleted {} stop times for pattern {}", deletedStopTimes, keyValue);
+                        // FIXME SQL injection
```
Comment (evansiroky) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Please make a prepared statement to avoid the possibility for SQL injection.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 4]
```diff
@@ -1425,12 +1418,14 @@ private static String getUpdateReferencesSql(
             case DELETE:
                 if (isArrayField) {
                     return String.format(
+                        // FIXME SQL injection
```
Comment (evansiroky) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Please make a prepared statement to avoid the possibility for SQL injection.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 5]
```diff
@@ -1425,12 +1418,14 @@ private static String getUpdateReferencesSql(
...
                         "delete from %s where %s @> ARRAY['%s']::text[]",
                         refTableName,
                         keyField.name,
                         keyValue
                     );
                 } else {
+                    // FIXME SQL injection
```
Comment (evansiroky) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Please make a prepared statement to avoid the possibility for SQL injection.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 6]
```diff
@@ -1439,6 +1434,7 @@ private static String getUpdateReferencesSql(
                     // replace the old value with the new value using array contains clause.
                     // FIXME This is probably horribly postgres specific.
                     return String.format(
+                        // FIXME SQL injection
```
Comment (evansiroky) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Please make a prepared statement to avoid the possibility for SQL injection.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 7]
```diff
@@ -1450,6 +1446,7 @@ private static String getUpdateReferencesSql(
                     );
                 } else {
                     return String.format(
+                        // FIXME SQL injection
```
Comment (evansiroky) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Please make a prepared statement to avoid the possibility for SQL injection.

**Comment by landonreed** [comment]
@evansiroky, I don't think the fact that this PR identifies these SQL injection possibilities warrants blocking this from being merged. Those FIXMEs are unrelated to the code I was working on -- I just happened to spot them while debugging. The appropriate thing to do would be to just add them to a new ticket and keep this PR focused on a single issue.

**Comment by evansiroky** [comment]
I'm gonna be stubborn again and not approve this PR until the SQL injection issues are fixed. SQL injection is a serious vulnerability and has consistently been ranked as the [#1 security vulnerability in software](https://www.owasp.org/images/7/72/OWASP_Top_10-2017_%28en%29.pdf.pdf). 

Furthermore, it is good practice to fix bugs as soon as you find them. You can read [numerous](https://www.ministryoftesting.com/dojo/lessons/ten-reasons-why-you-fix-bugs-as-soon-as-you-find-them) [articles](https://featureupvote.com/blog/bug-fixes-or-new-product-features/) dating back to the infamous [Joel Test](https://www.joelonsoftware.com/2000/08/09/the-joel-test-12-steps-to-better-code/) in the year 2000 that advise software development teams to fix bugs before making new features. Our minds are still fresh with this bit of code, so it will save time to do this now rather than later.

**Comment by abyrd** [comment]
I've got to agree with @evansiroky on this - SQL injection is a major problem and has to be avoided systematically. The only reason it was left as a fixme / todo early on in the development process was that we wanted to adopt a single uniform approach to avoiding it. I know this application is always running in a controlled environment, but if you've ever dug through the logs of a server on the public internet it's astounding to see how many break-in attempts there are, at the very least probes for weaknesses. People will attack anything to use it as a spam relay or DDOS node or just to trash data for fun.

I think the best defense is to tightly restrict what can appear in substituted strings (ideally only letters and underscores). Prepared statements are necessary in all other cases, and even desirable in the "safe" cases.

I know it's annoying but it's an inherent problem with sending free-text queries to a separate server process. In my opinion it's a fundamental weakness in the SQL database server model, but that's the price we pay for using long-lived, heavily used technologies.
